javalib/src/main/scala/java/util/concurrent/ConcurrentSkipListSet.scala

```diff
@@ -0,0 +1,404 @@
+    extends AbstractSet[E]
+    with NavigableSet[E]
+    with Cloneable
+    with Serializable { self =>
+
+  def this() =
+    this(ConcurrentSkipListSet.defaultOrdering[E], false)
+
+  def this(comparator: Comparator[E])  = 
+    this(Ordering.comparatorToOrdering(comparator), true)
+
+  def this(sortedSet: SortedSet[E]) = {
+  	this(Ordering.comparatorToOrdering(sortedSet.comparator()), true)
+    addAll(sortedSet)
+  }
+
+  def this(collection: java.util.Collection[E]) = {
+  	this(ConcurrentSkipListSet.defaultOrdering[E], false)
+    addAll(collection)
+  }
+
+  def this(comparator: Comparator[E], collection: Collection[E]) = {
+    this(Ordering.comparatorToOrdering(comparator), true)
+    addAll(collection)
+  }
+
+  implicit object BoxOrdering extends Ordering[Box[E]] {
+    def compare(a: Box[E], b:Box[E]): Int = ordering.compare(a.inner, b.inner)
+  }
+
+  private val inner: mutable.SortedSet[Box[E]] = new mutable.TreeSet[Box[E]]()
+
+  override def isEmpty(): Boolean =   
+    inner.isEmpty
+
+  def iterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.iterator
+
+      var actual: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        actual = Some(iter.next().inner)
+        actual.get
+      }
+
+      def remove(): Unit = actual.map(self.remove(_))
+    }
+  }
+
+  override def add(e: E): Boolean =
+    if (e === null) throw new NullPointerException()
+    else inner.add(Box(e))
+
+  override def addAll(c: Collection[_ <: E]): Boolean = {
+    val iter = c.iterator()
+    var changed = false
+    while (iter.hasNext)
+      changed = add(iter.next()) || changed
+    changed
+  }
+
+  override def clear(): Unit =
+    inner.clear() 
+
+  override def clone(): ConcurrentSkipListSet[E] =
+    new ConcurrentSkipListSet(this)
+
+  def comparator(): Comparator[E] = {
+    if (!provided) null
+    else  {
+      new Comparator[E] {
+        def compare(a: E, b: E) = ordering.compare(a, b)
+      }
+    }
+  }
+ 
+  override def contains(o: Any): Boolean =
+    inner.contains(Box(o.asInstanceOf[E]))
+
+  def descendingIterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.iterator.toIndexedSeq.reverse.iterator
+
+      var actual: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        actual = Some(iter.next().inner)
+        actual.get
+      }
+
+      def remove(): Unit = actual.map(self.remove(_))
+    }
+  }
+
+  def descendingSet(): NavigableSet[E] = {
+    val descSetFun = () => {
+      val inverseOrdering = 
+        new Ordering[Box[E]] {
+         def compare(a: Box[E], b: Box[E]) = self.ordering.compare(b.inner,a.inner)
+        }
+      val retSet = new mutable.TreeSet[Box[E]]()(inverseOrdering)
+      retSet.addAll(inner)
+      retSet
+    }
+    new ConcurrentSkipListSet.NavigableView(this, descSetFun)
```

Comment:
This will reexecute `descSetFun` for every single operation in `NavigableView`, and thus recreate a new `mutable.TreeSet` every single time. That does not look good.